src/Numerics/Mesh/Grids.jl

```diff
@@ -570,106 +591,242 @@ function commmapping(N, commelems, commfaces, nabrtocomm)
+
+
+    # Sort out the vgeo terms
+    @views begin
+        vgeo_N1_flds =
+            ntuple(fld -> reshape(vgeo_N1[:, fld, :], Nq_N1..., nelem), _nvgeo)
+
+        # Allocate the storage for N = 0 volume metrics
+        vgeo = zeros(FT, Np, _nvgeo, nelem)
+
+        # Counter to make sure we got all the vgeo terms
+        num_vgeo_handled = 0
+
+        # _M should be a sum
+        vgeo[:, _M, :][:] .= sum(vgeo_N1_flds[_M], dims = findall(Nq .== 1))[:]
+        num_vgeo_handled += 1
+
+        # need to recompute _MI
+        vgeo[:, _MI, :] = 1 ./ vgeo[:, _M, :]
+        num_vgeo_handled += 1
+
+        # coordinates should just be averages
+        avg_den = 2 .^ sum(Nq .== 1)
+        for fld in (_x1, _x2, _x3)
+            vgeo[:, fld, :] =
+                sum(vgeo_N1_flds[fld], dims = findall(Nq .== 1))[:] / avg_den
+            num_vgeo_handled += 1
+        end
+
+        # For the metrics it is J * ξixk we approximate so multiply and divide the
+        # mass matrix (which has the Jacobian determinant and the proper averaging
+        # due to the quadrature weights)
+        M_N1 = vgeo_N1_flds[_M]
+        MI = vgeo[:, _MI, :]
+        for fld in
+            (_ξ1x1, _ξ2x1, _ξ3x1, _ξ1x2, _ξ2x2, _ξ3x2, _ξ1x3, _ξ2x3, _ξ3x3)
+            vgeo[:, fld, :] =
+                sum(M_N1 .* vgeo_N1_flds[fld], dims = findall(Nq .== 1))[:] .*
+                MI[:]
+            num_vgeo_handled += 1
+        end
+
+        # compute MH and JvC
+        horizontal_vertical_metrics(vgeo, Nq, ω)
+        num_vgeo_handled += 2
+
+        # Make sure we handled all the vgeo terms
+        @assert _nvgeo == num_vgeo_handled
+    end
+
+    # Sort out the sgeo terms
+    @views begin
+        sgeo = zeros(FT, _nsgeo, maximum(Nfp), nface, nelem)
+
+        # for the volume inverse mass matrix
+        MI = vgeo[:, _MI, :]
+        p = reshape(1:Np, Nq)
+        if dim == 1
+            fmask = (p[1:1], p[Nq[1]:Nq[1]])
+        elseif dim == 2
+            fmask = (p[1, :][:], p[Nq[1], :][:], p[:, 1][:], p[:, Nq[2]][:])
+        elseif dim == 3
+            fmask = (
+                p[1, :, :][:],
+                p[Nq[1], :, :][:],
+                p[:, 1, :][:],
+                p[:, Nq[2], :][:],
+                p[:, :, 1][:],
+                p[:, :, Nq[3]][:],
+            )
+        end
+        for d in 1:dim
+            for f in (2d - 1):(2d)
+                # number of points matches means that we keep all the data
+                # (N = 0 is not on the face)
+                if Nfp[d] == Nfp_N1[d]
+                    sgeo[:, 1:Nfp[d], f, :] .= sgeo_N1[:, 1:Nfp[d], f, :]
+
+                    # Volume inverse mass will be wrong so reset it
+                    sgeo[_vMI, 1:Nfp[d], f, :] .= MI[fmask[f], :]
+                else
+                    # Counter to make sure we got all the sgeo terms
+                    num_sgeo_handled = 0
+
+                    # sum to get sM
```

Comment:
What are sM and normal.

are they related to  n ds = det(J) * J^-T * N dS,  J =∂x/∂ξ and N is the outward normal in the reference element? 

Is it because det(J) is linear and J^-T and N are constant with respect to the p0 direction coordinate, we can do summation ?

Reply:
> What are sM and normal.

`sM` is the surface mass matrix (with the surface jacobians scaled in), e.g., the sum off all these terms on a single face is the surface area of the face.

`normal` is the outward unit normal to a face in the physical space (the reference outward normal doesn't need to be stored)

> are they related to n ds = det(J) * J^-T * N dS, J =∂x/∂ξ and N is the outward normal in the reference element?

Yes. The normal is `n` and `ds` is almost `sM` except it includes the quadrature weights as we..

> Is it because det(J) is linear and J^-T and N are constant with respect to the p0 direction coordinate, we can do summation ?

`det(J)` isn't necessarily linear, but this is what you need to do so that the volume and surface metrics are consistently calculated (which is needed for conservation and constant preserving).